### PR TITLE
Location filter support 

### PR DESCRIPTION
**Pull request by ronpal**
# Description

Adding support for location filters

## Checklist

- [x] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [x] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [x] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [x] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            resource_cls=LocationFilter,
+            items=location_filter,
+            input_resource_cls=LocationFilterWrite,
+        )
+
+    def retrieve(self, id: int | Sequence[int]) -> LocationFilterList:
+        """Retrieve a LocationFilter.
+
+        Args:
+            data_set_id: Data set id of the LocationFilter.
```
[the remark this answers is not in the document]
Comment (ronpal) [reply cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:70]: This is just residue from copied code, will disappear 😇

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        All = AllScope
+        SpaceID = IDScope
+
+
+class LocationFilterCore(WriteableCogniteResource["LocationFilterWrite"], ABC):
+    """
+    LocationFilter contains information for a single LocationFilter.
+
+    Args:
+        externalId: The external ID provided by the client. Must be unique for the resource type.
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:114]:
```suggestion
        external_id: The external ID provided by the client. Must be unique for the resource type.
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    WriteableCogniteResource,
+    WriteableCogniteResourceList,
+)
+from cognite.client.data_classes.capabilities import AllScope, Capability, IDScope
+from cognite.client.data_classes.data_modeling.ids import DataModelId, ViewId
+from typing_extensions import Self
+
+
+@dataclass(frozen=True)
+class LocationFilterScene(DataModelId):
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:23]: Why are you subclassing `DataModelI`? LocationFilterScene does not seem to be a type of data model id

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+@dataclass(frozen=True)
+class LocationFilterAssetCentric:
+    assets: LocationFilterAssetCentricBaseFilter | None = None
+    events: LocationFilterAssetCentricBaseFilter | None = None
+    files: LocationFilterAssetCentricBaseFilter | None = None
+    timeseries: LocationFilterAssetCentricBaseFilter | None = None
+    sequences: LocationFilterAssetCentricBaseFilter | None = None
+    data_set_ids: list[int] | None = None
+    asset_subtree_ids: list[dict[str, int | str]] | None = None
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:66]: Hmm, maybe worth an special class her with identifier?
Alternative use a `Literal` .
```suggestion
    asset_subtree_ids: list[dict[Literal["id", "externalId"], int | str]] | None = None
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+@dataclass(frozen=True)
+class LocationFilterView(ViewId):
+    external_id: str
+    space: str
+    version: str | None = None
+    represents_entity: Literal["MAINTENANCE_ORDER", "OPERATION", "NOTIFICATION", "ASSET"] | None = None
+
+
+@dataclass(frozen=True)
+class LocationFilterAssetCentricBaseFilter:
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:44]: Filter twice in the name, what about?
All nested classes of a `CogniteResource` should use `CogniteObject`
```suggestion
class AssetCentricSubFilter(CogniteObject):
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    data_set_ids: list[int] | None = None
+    asset_subtree_ids: list[dict[str, str]] | None = None
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:46]: I am in favor of introducing a small class her Identifier. 
```suggestion
    asset_subtree_ids: list[dict[Literal["id", "externalId"], int | str]] | None = None
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    external_id_prefix: str | None = None
+
+    @classmethod
+    def load(cls, resource: dict[str, Any]) -> Self:
+        return cls(
+            data_set_ids=resource.get("dataSetIds"),
+            asset_subtree_ids=resource.get("assetSubtreeIds"),
+            external_id_prefix=resource.get("externalIdPrefix"),
+        )
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:55]: Subclassing `CogniteObject` gives you this one for free.
```suggestion
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    sequences: LocationFilterAssetCentricBaseFilter | None = None
+    data_set_ids: list[int] | None = None
+    asset_subtree_ids: list[dict[str, int | str]] | None = None
+    external_id_prefix: str | None = None
+
+    @classmethod
+    def load(cls, resource: dict[str, Any]) -> Self:
+        return cls(
+            assets=LocationFilterAssetCentricBaseFilter.load(resource.get("assets", {}))
+            if resource.get("assets")
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:73]: Times 5
```suggestion
            if "asset" in resource
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+    external_id: str
+    space: str
+    version: str
+
+
+@dataclass(frozen=True)
+class LocationFilterView(ViewId):
+    external_id: str
+    space: str
+    version: str | None = None
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:39]: No need to repeat parameters.
```suggestion
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+class LocationFilterScene(DataModelId):
+    external_id: str
+    space: str
+
+
+@dataclass(frozen=True)
+class LocationFilterDataModel(DataModelId):
+    external_id: str
+    space: str
+    version: str
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:32]: Is this to have required `version`? I suggest making a new class and not sublcass DataModelId then.

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+        )
+
+
+class LocationFilter(LocationFilterCore):
+    """
+    LocationFilter contains information for a single LocationFilter
+
+    Args:
+        id: The ID of the location filter
+        externalId: The external ID provided by the client. Must be unique for the resource type.
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:191]:
```suggestion
        external_id: The external ID provided by the client. Must be unique for the resource type.
```

File: cognite_toolkit/_cdf_tk/client/data_classes/locations.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    def load(cls, resource: dict[str, Any]) -> Self:
+        return cls(
+            data_set_ids=resource.get("dataSetIds"),
+            asset_subtree_ids=resource.get("assetSubtreeIds"),
+            external_id_prefix=resource.get("externalIdPrefix"),
+        )
+
+
+@dataclass(frozen=True)
+class LocationFilterAssetCentric:
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/data_classes/locations.py:59]: What about?
```suggestion
class AssetCentricFilter:
```

File: cognite_toolkit/_cdf_tk/client/api/locations/api.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+from cognite.client import ClientConfig, CogniteClient
+from cognite.client._api_client import APIClient
+
+from cognite_toolkit._cdf_tk.client.api.locations.location_filters import LocationFiltersAPI
+
+
+class LocationsAPI(APIClient):
+    def __init__(self, config: ClientConfig, api_version: str | None, cognite_client: CogniteClient) -> None:
+        super().__init__(config, api_version, cognite_client)
+        self.location_filters = LocationFiltersAPI(config, api_version, cognite_client)
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/api.py:12]: This becomes `client.locations.locations_filters` I think `client.locations.filters` is better.
```suggestion
        self.filters = LocationFiltersAPI(config, api_version, cognite_client)
```

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+    @overload
+    def create(self, location_filter: Sequence[LocationFilterWrite]) -> LocationFilterList: ...
+
+    def create(
+        self, location_filter: LocationFilterWrite | Sequence[LocationFilterWrite]
+    ) -> LocationFilter | LocationFilterList:
+        """Create a new LocationFilter.
+
+        Args:
+            LocationFilter: LocationFilterWrite or list of LocationFilterWrite.
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:64]:
```suggestion
            location_filter: LocationFilterWrite or list of LocationFilterWrite.
```

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+
+        return LocationFilterList(
+            [
+                LocationFilter._load(
+                    self._post(url_path=self._RESOURCE_PATH, json=item.dump()).json(),
+                    cognite_client=self._cognite_client,
+                )
+                for item in location_filter
+            ]
+        )
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:83]: I think the client should not invent to much, lets do this in the loader instead.

If the endpoint only supports creating one location at a time, then the client does the same

```suggestion
        res = self._post(url_path=self._RESOURCE_PATH, json=location_filter.dump())
        return LocationFilter._load(res.json(), cognite_client=self._cognite_client)
```

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+
+    def retrieve(self, external_id: str) -> LocationFilter | None:
+        for loc in self.list():
+            if loc.external_id == external_id:
+                return loc
+        raise CogniteNotFoundError(not_found=[external_id])
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:89]: Move to loader
```suggestion
```

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+        """
+        return iter(self.list())
+
+    def __iter__(self) -> Iterator[LocationFilter]:
+        return self.__call__()
+
+    @property
+    @lru_cache(maxsize=1)
+    def _ids(self) -> dict[str, int]:
+        return {loc.external_id: loc.id for loc in self.list()}
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:50]: Move to loader

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py (hunk 16 above)
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:89]: Why not implement the actual retrieve endpoint instead?
image

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+
+        return LocationFilterList(
+            [
+                LocationFilter._load(
+                    self._put(url_path=f"{self._RESOURCE_PATH}/{self._ids[item.external_id]}", json=item.dump()).json(),
+                    cognite_client=self._cognite_client,
+                )
+                for item in location_filter
+            ]
+        )
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:115]: Again move support for multiple locations to loader.

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+
+    def delete(self, ids: int | Sequence[int]) -> int:
+        if isinstance(ids, int):
+            ids = [ids]
+
+        n = 0
+        for id in ids:
+            self._delete(url_path=f"{self._RESOURCE_PATH}/{id}")
+            n += 1
+        return n
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:125]: Same here

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+
+    def list(self) -> LocationFilterList:
+        res = self._post(url_path=self._RESOURCE_PATH + "/list", json={"flat": False})
+        return LocationFilterList._load(res.json()["items"], cognite_client=self._cognite_client)
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:129]: There is a method `self._list(...)` that typically we use. It is not necessary as there is no paging for this endpoint, but I would requests that from the backend team

File: cognite_toolkit/_cdf_tk/loaders/_resource_loaders/location_loaders.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+        return item.external_id
+
+    def load_resource(
+        self, filepath: Path, ToolGlobals: CDFToolConfig, skip_validation: bool
+    ) -> LocationFilterWrite | LocationFilterWriteList:
+        resources = load_yaml_inject_variables(filepath, {})
+        if isinstance(resources, dict):
+            return LocationFilterWrite.load(resources)
+        else:
+            return LocationFilterWriteList.load(resources)
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/loaders/_resource_loaders/location_loaders.py:60]: This is the same as the parent class, so no need to reimplement it.
```suggestion
```

However, you need to implement it and deal with all the `dataSetId` to `dataSetExternalId` replacements.

File: cognite_toolkit/_cdf_tk/loaders/_resource_loaders/location_loaders.py [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+        if not items:
+            return []
+        return LocationFilterAcl(
+            actions=[LocationFilterAcl.Action.Read, LocationFilterAcl.Action.Write],
+            scope=LocationFilterAcl.Scope.All(),
+            allow_unknown=True,
+        )
+
+    @classmethod
+    def get_id(self, item: LocationFilter | LocationFilterWrite | dict) -> str:
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/loaders/_resource_loaders/location_loaders.py:46]:
```suggestion
    def get_id(cls, item: LocationFilter | LocationFilterWrite | dict) -> str:
```

File: cognite_toolkit/_cdf_tk/loaders/_resource_loaders/location_loaders.py [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+
+    def update(self, items: LocationFilterWriteList) -> LocationFilterList:
+        return self.client.locations.location_filters.update(items)
+
+    def delete(self, external_ids: SequenceNotStr[str]) -> int:
+        ids = [loc.id for loc in self.retrieve(external_ids)]
+        return self.client.locations.location_filters.delete(ids)
+
+    def iterate(self) -> Iterable[LocationFilter]:
+        return iter(self.client.locations.location_filters)
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/loaders/_resource_loaders/location_loaders.py:78]: In addition, you need to implement `get_write_cls_parameter_spec` and fix all the datasetID issues in the AssetCentric part of the AssetCentric.

File: tests/test_unit/approval_client/config.py [hunk 24]
```diff
@@ -84,11 +84,6 @@
     WorkflowVersionUpsert,
     WorkflowVersionUpsertList,
 )
-
```
Comment (doctrino) [inline tests/test_unit/approval_client/config.py:87]: This needs to be extended.

**Review by doctrino — CHANGES_REQUESTED** [review]
Main feedback it to move the patching of the endpoint into the loader and not the client.

In addition, need to update the config for the ApprovalClient as well as implementing the `get_write_cls_parameter_spec` method in thhe loader.

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        return iter(self.list())
+
+    def __iter__(self) -> Iterator[LocationFilter]:
+        return self.__call__()
+
+    # @property
+    # @lru_cache(maxsize=1)
+    # def _ids(self) -> dict[str, int]:
+    #     return {loc.external_id: loc.id for loc in self.list()}
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:48]:
```suggestion
```

File: cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+        Yields:
+            LocationFilter or LocationFilterList
+
+        """
+        return iter(self.list())
+
+    def __iter__(self) -> Iterator[LocationFilter]:
+        return self.__call__()
+
+
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/client/api/locations/location_filters.py:45]:
```suggestion
```

**Review by doctrino — APPROVED** [review]
Finished up the PR.

**Pull request merged**